src/Sentry/SentryClient.cs

```diff
@@ -281,6 +267,34 @@ private SentryId DoSendEvent(SentryEvent @event, Hint? hint, Scope? scope)
             return SentryId.Empty;
         }
 
+        var hasTerminalException = processedEvent.HasTerminalException();
+        if (hasTerminalException)
+        {
+            // Event contains a terminal exception -> end session as crashed
+            _options.LogDebug("Ending session as Crashed, due to unhandled exception.");
+            scope.SessionUpdate = _sessionManager.EndSession(SessionEndStatus.Crashed);
+        }
+        else if (processedEvent.HasException())
+        {
+            // Event contains a non-terminal exception -> report error
+            // (this might return null if the session has already reported errors before)
+            scope.SessionUpdate = _sessionManager.ReportError();
+        }
+
+        if (_options.SampleRate != null)
+        {
+            if (!_randomValuesFactory.NextBool(_options.SampleRate.Value))
+            {
+                _options.ClientReportRecorder.RecordDiscardedEvent(DiscardReason.SampleRate, DataCategory.Error);
+                _options.LogDebug("Event sampled.");
+                return SentryId.Empty;
+            }
+        }
+        else
+        {
+            _options.LogDebug("Event not sampled.");
+        }
```

Comment:
```suggestion
```
Alternatively, this block could be moved one line up?

Reply:
It's mainly there for testing... so that we can ensure the order of processing when the sample rate is null. 

The problem with putting it in the block where the sample rate is not null is that it would make the tests non-deterministic (you're rolling dice to see whether the tests pass then). 

Or did you have something else in mind?